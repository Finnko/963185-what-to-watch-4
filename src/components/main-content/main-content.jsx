import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import MoviePropType from '../../prop-types/movie';
import {Config, ViewMode} from '../../const';
import {ActionCreator} from '../../store/actions/action-creator';
import {getActiveGenre, getGenres, getMoviesByGenre, getMoviesRenderLimit} from '../../store/reducers/app/selectors';
import MoviesList from '../movies-list/movies-list.jsx';
import ShowMore from '../show-more/show-more.jsx';
import GenresList from '../genres-list/genres-list.jsx';
import withActiveItem from '../../hocs/with-active-item/with-active-item';

const GenresListWrapped = withActiveItem(GenresList);

const MainContent = ({
  genres,
  filteredMovies,
  activeGenre,
  renderLimit,
  handleGenreChange,
  handleShowMoreClick
}) => {
  const isShowMoreVisible = renderLimit > filteredMovies.length;

  return (
    <section className="catalog">
      <h2 className="catalog__title visually-hidden">Catalog</h2>

      <GenresListWrapped
        activeItem={activeGenre}
        genres={genres}
        onGenreChange={handleGenreChange}
      />

      <MoviesList
        movies={filteredMovies}
        viewMode={ViewMode.MOVIE_CARD.WITH_PLAYER}
      />

      {!isShowMoreVisible &&
        <ShowMore
          offset={Config.MOVIES_NUMBER_OFFSET}
          onButtonClick={handleShowMoreClick}
        />
      }
    </section>
  );
};

MainContent.propTypes = {
  filteredMovies: PropTypes.arrayOf(MoviePropType).isRequired,
  activeGenre: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  renderLimit: PropTypes.number,
  handleGenreChange: PropTypes.func.isRequired,
  handleShowMoreClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  return {
    filteredMovies: getMoviesByGenre(state),
    activeGenre: getActiveGenre(state),
    genres: getGenres(state),
    renderLimit: getMoviesRenderLimit(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  handleGenreChange(genre) {
    dispatch(ActionCreator.changeActiveGenre(genre));
    dispatch(ActionCreator.resetMoviesLimit());
  },
  handleShowMoreClick(offset) {
    dispatch(ActionCreator.changeMoviesLimit(offset));
  }
});

export {MainContent};
export default connect(mapStateToProps, mapDispatchToProps)(MainContent);

