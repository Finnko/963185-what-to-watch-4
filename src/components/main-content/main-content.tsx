import * as React from 'react';
import {connect} from 'react-redux';
import {Config, ViewMode} from '../../const';
import {Movie} from '../../interfaces';
import {ActionCreator} from '../../store/actions/action-creator';
import {getActiveGenre, getGenres, getMoviesByGenre, getMoviesRenderLimit} from '../../store/reducers/app/selectors';
import MoviesList from '../movies-list/movies-list';
import ShowMore from '../show-more/show-more';
import GenresList from '../genres-list/genres-list';
import withActiveItem from '../../hocs/with-active-item/with-active-item';

const GenresListWrapped = withActiveItem(GenresList);

type Props = {
  genres: string[],
  filteredMovies: Movie[],
  activeGenre: string,
  renderLimit: number,
  handleGenreChange: () => void,
  handleShowMoreClick: () => void,
}

const MainContent:React.FC<Props> = ({
  genres,
  filteredMovies,
  activeGenre,
  renderLimit,
  handleGenreChange,
  handleShowMoreClick
}) => {
  const isShowMoreVisible = renderLimit > filteredMovies.length;

  return (
    <section className='catalog'>
      <h2 className='catalog__title visually-hidden'>Catalog</h2>

      <GenresListWrapped
        activeItem={activeGenre}
        genres={genres}
        onGenreChange={handleGenreChange}
      />

      <MoviesList
        movies={filteredMovies}
        viewMode={ViewMode.MOVIE_CARD.WITH_PLAYER}
      />

      {!isShowMoreVisible &&
        <ShowMore
          offset={Config.MOVIES_NUMBER_OFFSET}
          onButtonClick={handleShowMoreClick}
        />
      }
    </section>
  );
};

const mapStateToProps = (state) => {
  return {
    filteredMovies: getMoviesByGenre(state),
    activeGenre: getActiveGenre(state),
    genres: getGenres(state),
    renderLimit: getMoviesRenderLimit(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  handleGenreChange(genre) {
    dispatch(ActionCreator.changeActiveGenre(genre));
    dispatch(ActionCreator.resetMoviesLimit());
  },
  handleShowMoreClick(offset) {
    dispatch(ActionCreator.changeMoviesLimit(offset));
  }
});

export {MainContent};
export default connect(mapStateToProps, mapDispatchToProps)(MainContent);

