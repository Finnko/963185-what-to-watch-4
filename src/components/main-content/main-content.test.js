import React from 'react';
import renderer from 'react-test-renderer';
import {createMemoryHistory} from 'history';
import {Router} from 'react-router-dom';
import configureStore from 'redux-mock-store';
import NameSpace from '../../store/name-space';
import {Provider} from 'react-redux';
import {MainContent} from './main-content';

const mock = [
  {
    id: `1`,
    title: `The Grand Budapest Hotel`,
    genre: `Drama`,
    preview: `img/bohemian-rhapsody.jpg`,
    releaseYear: 2014,
    videoSrc: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    backgroundImage: `img/bg-the-grand-budapest-hotel.jpg`,
    description: [
      `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
      `Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`
    ],
    director: `Michael Bay`,
    actors: [`Leonardo Di Caprio`],
    rating: 7.5,
    ratingCount: 250,
    poster: `/img/the-grand-budapest-hotel-poster.jpg`,
  }
];

const mockStore = configureStore([]);

const store = mockStore({
  [NameSpace.DATA]: {
    movies: mock,
  },
  [NameSpace.APP]: {
    activeGenre: `All genres`,
    renderLimit: 8,
  }
});

const props = {
  filteredMovies: mock,
  activeGenre: `All genres`,
  renderLimit: 8,
  genres: [`All genres`, `Crime`, `Horror`],
  handleGenreChange: jest.fn(),
  handleShowMoreClick: jest.fn(),
};

describe(`MainContent component render correctly`, () => {
  it(`Should MainContent component render correctly`, () => {
    const history = createMemoryHistory();

    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <MainContent {...props}/>
            </Router>
          </Provider>, {
            createNodeMock: () => {
              return {};
            },
          }
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
