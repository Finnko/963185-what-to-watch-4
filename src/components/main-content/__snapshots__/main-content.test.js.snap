// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContent component render correctly Should MainContent component render correctly 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Crime
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Horror
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        className="small-movie-card__link"
        href="/movie/1"
        onClick={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <video
            height="175"
            muted={true}
            poster="img/bohemian-rhapsody.jpg"
            src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
            width="100%"
          />
        </div>
      </a>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/movie/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
  </div>
</section>
`;
